src/Core/Grid/Data/Factory/DoctrineGridDataFactory.php

```diff
@@ -93,9 +95,39 @@ public function getData(SearchCriteriaInterface $searchCriteria)
             'search_criteria' => $searchCriteria,
         ]);
 
-        $records = $searchQueryBuilder->execute()->fetchAll();
         $recordsTotal = (int) $countQueryBuilder->execute()->fetch(PDO::FETCH_COLUMN);
+        $searchOffset = (int) $searchCriteria->getOffset();
+
+        // If offset is out of range, we compute and apply a valid offset.
+        if (Pagination::isOffsetOutOfRange($recordsTotal, $searchOffset)) {
+            // @TODO : those cases should be simplified with some refactoring cf. https://github.com/PrestaShop/PrestaShop/issues/29131
+
+            $searchLimit = $searchCriteria->getLimit();
+            $inRangeOffset = Pagination::computeValidOffset($recordsTotal, $searchOffset, $searchLimit);
+
+            // @TODO : add INFO log to prevent that we overwrite offset from `$searchCriteria->getOffset()` to `$inRangeOffset`
 
+            if ($searchCriteria instanceof Filters) {
+                $newSearchCriteria = clone $searchCriteria;
+                $newSearchCriteria->add(['offset' => $inRangeOffset]);
```

Comment:
Just a question here why don't we use Filters::addFilter() ?

Reply:
> Because it's not a filters that we add here, it's one of the other options stored in the Filters data on an upper level. That's the structure of all the data:
> 
> ```
>         [
>             'limit' => 50,
>             'offset' => 0,
>             'orderBy' => 'id_address',
>             'sortOrder' => 'asc',
>             'filters' => [],
>         ]
> ```
> 
> We want to change `offset` and `addFilters` only changes the `filters` part

I didn't see your comment @jolelievre , you are right ! I missed the key  xD